### PR TITLE
add support for diffing multiple simulation screenshots

### PR DESCRIPTION
**Pull request by alexivanov**
### Changes

We're adding the ability to take multiple screenshots during simulations.

This PR extends the existing diffing utilities to handle comparing  simulations with multiple screenshots.

In cases where we have multiple screenshots diffing can "fail" if any of the following is true:
 * for any screenshot pair, the pixel mismatch is above the set threshold
 * overall pixel mismatch across all screenshot pairs is above the set threshold
 * the head simulation is missing a screenshot which is present in the base simulation-- currently we assume the base simulation screenshots are a subset of the the head simulation.

The failing criteria can change in the future but this preserves the current behaviour (as `final-state` screenshots is a special case of the multiple screenshot simulations)

### CODE REVIEW

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 1]
```diff
@@ -23,79 +26,198 @@ export class DiffError extends Error {
...
-  headScreenshot,
+  baseScreenshotsDir,
+  headScreenshotsDir,
   threshold: threshold_,
   pixelThreshold,
   exitOnMismatch,
 }) => {
   const logger = log.getLogger(METICULOUS_LOGGER_NAME);
 
   const threshold = threshold_ || DEFAULT_MISMATCH_THRESHOLD;
```
Comment (leafty) [inline packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:67]: I think this is a mistake from me:

```suggestion
  const threshold = threshold_ ?? DEFAULT_MISMATCH_THRESHOLD;
```

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 2]
```diff
@@ -23,79 +26,198 @@ export class DiffError extends Error {
...
+          baseReplayId,
+          headReplayId,
+          threshold,
+        }
+      );
+    }
+
+    const totalMismatchedFraction = totalMismatchPixels / totalComparedPixels;
+    const overallMismatchOutcome: ComparisonOutcome =
+      totalMismatchedFraction > threshold ? "fail" : "pass";
```
Comment (leafty) [inline packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:199]: I think doing the overall check is not needed (it's impossible to fail).

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 3]
```diff
@@ -111,6 +233,16 @@ export const diffScreenshots: (options: {
...
+
+  return comparisonResults;
+};
+
+const logComparisonResultMessage: (
+  logger: Logger,
+  message: string,
+  outcome: ComparisonOutcome
+) => void = (logger, message, outcome) => {
+  logger.info(`${message} => ${outcome === "fail" ? "FAIL!" : "PASS"}`);
```
Comment (leafty) [inline packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:245]: nitpick: more foolproof if there are more outcome types later.

```suggestion
  logger.info(`${message} => ${outcome === "pass" ? "PASS" : "FAIL!"}`);
```

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts (hunk 2 above)
Comment (alexivanov) [reply packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts]: ah yes, good point. AVG > threshold will be caught by the above individual screenshot checks https://github.com/alwaysmeticulous/meticulous-sdk/pull/132/commits/4ac49190537a6b06906cb85e0a2cf57b08e0b9c2

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 4]
```diff
@@ -23,79 +26,172 @@ export class DiffError extends Error {
...
+      comparisonResults.push({
+        baseScreenshotFile,
+        headScreenshotFile,
+        comparisonResult,
+        outcome:
+          comparisonResult.mismatchFraction > threshold ? "fail" : "pass",
+      });
+    }
 
     await postScreenshotDiffStats(client, {
```
[the remark this answers is not in the document]
Comment (leafty) [reply packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:146]: I think we can do that by updating the [TestRun](https://github.com/alwaysmeticulous/meticulous/blob/master/packages/database-entities/src/test-run/test-run.entity.ts) entity and processing on the backend, yes :+1:

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 5]
```diff
@@ -23,79 +26,172 @@ export class DiffError extends Error {
...
+      const headScreenshot = await readPng(headScreenshotFile);
+
+      const comparisonResult = compareImages({
+        base: baseScreenshot,
+        head: headScreenshot,
+        pixelThreshold: pixelThreshold ?? null,
+      });
+
+      totalMismatchPixels += comparisonResult.mismatchPixels;
+      totalComparedPixels += baseScreenshot.width * baseScreenshot.height;
```
Comment (leafty) [inline packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:122]: Reported not used

**Review by leafty — APPROVED** [review]
LGTM :clap:

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 6]
```diff
@@ -23,79 +26,172 @@ export class DiffError extends Error {
...
+  const threshold = threshold_ ?? DEFAULT_MISMATCH_THRESHOLD;
+
+  const baseReplayScreenshots = await getScreenshotFiles(baseScreenshotsDir);
+  const headReplayScreenshots = await getScreenshotFiles(headScreenshotsDir);
+
+  // Assume base replay screenshots are always a subset of the head replay screenshots.
+  // We report any missing base replay screenshots for visibility but don't count it as a difference.
+  const missingHeadImages = new Set(
+    [...baseReplayScreenshots].filter(
+      (file_) => !headReplayScreenshots.includes(file_)
```
Comment (leafty) [inline packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:76]: I forgot this one:

```suggestion
      (file) => !headReplayScreenshots.includes(file)
```

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts [hunk 7]
```diff
@@ -1,18 +1,21 @@
...
   getOrFetchReplayArchive,
-  readReplayScreenshot,
+  getReplayDir,
+  getScreenshotFiles,
+  getScreenshotsDir,
 } from "../../local-data/replays";
 import { writeScreenshotDiff } from "../../local-data/screenshot-diffs";
 import { wrapHandler } from "../../utils/sentry.utils";
+import { readPng } from "../../image/io.utils";
+import { basename, join } from "path";
```
Comment (leafty) [inline packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:18]: Nitpick: run "optimize imports"

**Review by leafty — COMMENTED** [review]
Forgot these comments

File: packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts (hunk 7 above)
Comment (alexivanov) [reply packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts:18]: My IDE thinks imports are ordered correctly here 🤔 wonder if should standardise import ordering with eslint / prettier?

Comment (alexivanov) [reply packages/cli/src/commands/screenshot-diff/screenshot-diff.command.ts]: https://github.com/alwaysmeticulous/meticulous-sdk/pull/132/commits/eff308933321daf8e9a050292f84bddfc83d3f3c

**Pull request merged**